### PR TITLE
WebAuthn on port 21326

### PR DESCRIPTION
**Pull request by andrewkozlik**
When running `emu.py` (debug) and `emu.py -d` (production), the WebAuthn UDP port is different in each case, 21326 and 21325, respectively. There are two problems:
1. This is an unwelcome complication when running `hid-bridge`, because you have to set the port manually.
2. `emu.py -d` shows the wrong port number:
```
TREZOR_PATH=udp:127.0.0.1:21324
TREZOR_PROFILE_DIR=/var/tmp
TREZOR_UDP_PORT=21324
TREZOR_FIDO2_UDP_PORT=21326  # <---- here
```

We can easily unify the WebAuthn port for both debug and production by reserving the second interface for debug. That way, if WebAuthn is enabled, it will always be on the third interface, meaning that by default the emulator will run WebAuthn on port 21326 every time (except for bitcoin-only).

### CODE REVIEW

**Comment by mcudev** [comment]
Thanks for this. I've reviewed your proposed changes and have pushed up something for you to consider at: https://github.com/trezor/trezor-firmware/compare/master...mcudev:webauthn_port_21325

Here is my review:

1. The `id_debug` variable is moved out of it's static branch and would no longer be excluded from the bytecode.
1. The endpoint addresses of the endpoint descriptors for the WebAuthn HID interface would be incremented by one more than current for production devices (currently, the wire interface gets 0x81 and 0x01; the webauthn / hid interface gets 0x82 and 0x02). Would that cause a problem for already installed devices (and on Windows)? I don't know, but should test. Also, would need to test and see if non-sequential endpoint addresses cause any device issues (I don't know, probably not, it's just a difference to test).
1. We can avoid the prior two things by moving the debug interfaces after the webauthn like is shown in my suggestion referenced above and use port **21325** for webauthn in the emulator. This seems to work for me and I don't see much potential downside (it's less production runtime difference).

**Comment by prusnak** [comment]
Yes, I'd rather shuffle with the IDs of the debug firmware rather than IDs of the production firmware for the reason above.

**Comment by mcudev** [comment]
Also, I just made one more small change in usb.py in my suggestion that adds the `iface_webauthn` before `iface_debug` similarly too: https://github.com/trezor/trezor-firmware/compare/master...mcudev:webauthn_port_21325

**Comment by prusnak** [comment]
> Also, I just made one more small change in usb.py in my suggestion that adds the `iface_webauthn` before `iface_debug` similarly too: [master...mcudev:webauthn_port_21325](https://github.com/trezor/trezor-firmware/compare/master...mcudev:webauthn_port_21325)

That's exactly the change I had in mind, thanks!

**Comment by prusnak** [comment]
One thing to consider: Although, this change seems straightforward (and it is), we need to evaluate how much we'd need to change our testing stack. The changes for the python test suite should be also very straightforward (not sure there are even some necessary). But there might be issues with the upgrade tests, where the old firmwares will have different layout of interfaces/endpoints (resp. UDP ports for emulator) and we might need to add some extra logic to handle both cases. 

I am paging @tsusanka and @matejcik as they know the most about python stuff + CI.

**Comment by andrewkozlik** [comment]
> we need to evaluate how much we'd need to change our testing stack

If we just switch the order of debuglink and WebAuthn interfaces, then the problem that this is trying to solve will shift from WebAuthn to debuglink. What I mean is that on the standard debug firmware you will have debuglink on port 21326 and on bitcoin-only debug firmware it will run on 21325. So I suppose we should first figure out whether non-sequential endpoint addresses cause any issues. Also, we will need to update the documentation, e.g. here https://github.com/trezor/trezord-go#debug-link-support.

**Comment by andrewkozlik** [comment]
FYI, I just tested this branch with the skipped debug endpoint on the actual device, i.e. non-debug non-bitcoin-only firmware, using Chromium on Ubuntu at https://webauthn.io/ and it works fine.

**Comment by prusnak** [comment]
> If we just switch the order of debuglink and WebAuthn interfaces, then the problem that this is trying to solve will shift from WebAuthn to debuglink.

Right, but possible issue with the debuglink interface affects only hundreds of (technically advanced) users, while potential issue with the WebAuthn interface affects all users out there.

**Comment by prusnak** [comment]
> FYI, I just tested this branch with the skipped debug endpoint on the actual device

I think that having non-sequential endpoints is fine, so this should not cause a problem itself. What we have to deal with is for example Windows which will try to download new drivers for new endpoints, which might cause unnecessary breakage.

What if we moved the debuglink to endpoints 83/03 and kept webauthn at 82/02 (with an optional skip for btc-only firmware)?

**Comment by mcudev** [comment]
I did some testing on Ubuntu 20.10. only the first config worked, the rest failed either on the TT device or on the host computer. It does not appear that gaps in the endpoint addresses work (unless I'm interpreting the data wrong somehow):

```
works, `screen /dev/ttyACM0` follows device output, `trezorctl debug show-text abc` works, host kernel messages look good:
wire  iface_num = 0, ep_in = 0x81, ep_out = 0x01
debug iface_num = 1, ep_in = 0x82, ep_out = 0x02
vcp   iface_num = 2, ep_in = 0x83, ep_out = 0x03, data_iface_num = 3, ep_cmd=0x84

does not work, TT device halts on the "Processing" screen, kernel logs "ttyACM0: USB ACM device" and "USB disconnect" when unplugged, seen by lsusb:
wire  iface_num = 0, ep_in = 0x81, ep_out = 0x01
debug iface_num = 1, ep_in = 0x83, ep_out = 0x03
vcp   iface_num = 2, ep_in = 0x84, ep_out = 0x04, data_iface_num = 3, ep_cmd=0x85

does not work, TT device halts on the "Processing" screen, kernel logs "config 1 has an invalid interface number: 4 but max is 3", "config 1 has no interface number 1", "ttyACM0: USB ACM device" and "USB disconnect" when unplugged, seen by lsusb:
wire  iface_num = 0, ep_in = 0x81, ep_out = 0x01
debug iface_num = 2, ep_in = 0x83, ep_out = 0x03
vcp   iface_num = 3, ep_in = 0x84, ep_out = 0x04, data_iface_num = 4, ep_cmd=0x85

does not work, TT device gets to home screen, causes kernel error dump messages when trying to use `screen /dev/ttyACM0`:
wire  iface_num = 0, ep_in = 0x81, ep_out = 0x01
debug iface_num = 1, ep_in = 0x87, ep_out = 0x07
vcp   iface_num = 2, ep_in = 0x88, ep_out = 0x08, data_iface_num = 3, ep_cmd=0x89
```

**Comment by mcudev** [comment]
here are some more tests on a TT device with a ubuntu 20.10 host. this batch were non-debug firmware builds.
i think my summary would be to not skip interface numbers and to avoid skipping endpoint addresses if possible.
the last failure case feels like it's hitting a bug in the TT USB stack somewhere, but I did not dig into it yet:

```
seems to work cleanly:
wire     iface_num = 0, ep_in = 0x81, ep_out = 0x01
webauthn iface_num = 1, ep_in = 0x82, ep_out = 0x02

seems to work cleanly:
wire     iface_num = 0, ep_in = 0x81, ep_out = 0x01
webauthn iface_num = 1, ep_in = 0x83, ep_out = 0x03

seems to work, but the kernel also logs this "config 1 has an invalid interface number: 2 but max is 1" and "config 1 has no interface number 1":
wire     iface_num = 0, ep_in = 0x81, ep_out = 0x01
webauthn iface_num = 2, ep_in = 0x83, ep_out = 0x03

device works until a webauthn is attempted then it seems to lockup (first `trezorctl list` command hangs and further return "LIBUSB_ERROR_BUSY"):
wire     iface_num = 0, ep_in = 0x81, ep_out = 0x01
webauthn iface_num = 1, ep_in = 0x84, ep_out = 0x04
```

**Comment by andrewkozlik** [comment]
I hadn't realized that skipping the interface numbers would be such a problem. In that case this is not a good approach. Since this is an emulator issue, it would be better not to touch the interface code above, but to go about this by making the emulator map WebAuthn to `TREZOR_UDP_PORT + 2` every time.

File: core/embed/trezorhal/usb_vcp-defs.h [hunk 1]
```diff
@@ -105,9 +105,10 @@ typedef struct {
   uint8_t iface_num;         // Address of this VCP interface
   uint8_t data_iface_num;    // Address of data interface of the VCP interface
                              // association
-  uint8_t ep_cmd;  // Address of IN CMD endpoint (with the highest bit set)
-  uint8_t ep_in;   // Address of IN endpoint (with the highest bit set)
-  uint8_t ep_out;  // Address of OUT endpoint
+  uint8_t ep_cmd;     // Address of IN CMD endpoint (with the highest bit set)
+  uint8_t ep_in;      // Address of IN endpoint (with the highest bit set)
+  uint8_t ep_out;     // Address of OUT endpoint
+  uint16_t emu_port;  // UDP port of this interface in the emulator.
```
Comment (prusnak) [inline core/embed/trezorhal/usb_vcp-defs.h:111]: Can we use maybe the following trick?

That way we would not need to modify unrelated code (bootloader/etc).

``` c
#ifdef TREZOR_EMULATOR
   uint16_t emu_port;  // UDP port of this interface in the emulator.
#else
  uint8_t ep_cmd;     // Address of IN CMD endpoint (with the highest bit set)
  uint8_t ep_in;      // Address of IN endpoint (with the highest bit set)
  uint8_t ep_out;     // Address of OUT endpoint
#endif
```

Comment (andrewkozlik) [reply core/embed/trezorhal/usb_vcp-defs.h]: https://github.com/trezor/trezor-firmware/pull/1418/commits/48f81ebca45b211b056b5cee00ab79538683995c

**Comment by mcudev** [comment]
I've been looking at the USB code today and wondering **why** we are encountering this endpoint count restriction hoping that we can use all of these interfaces at the same time)?

I see that we use the USB HS (high-speed) controller in FS (full-speed) mode (`USE_USB_HS` and `USE_USB_HS_IN_FS` are defined in `trezor-firmware/core/embed/trezorhal/stm32f4xx_hal_conf.h`).

But, **why** does HS in FS mode require not using all of the endpoints that the HS controller has available?

The OTG_HS peripheral in peripheral mode has 1 bidirectional control endpoint 0 (address 0x80) plus 5 IN endpoints plus 5 OUT endpoints (see section 35.5.3 and 35.2.3 in RM0090, the reference manual for the STM32F427, and AN4879).
I only see bidirectional control endpoint 0 being used in addition to the in and out endpoints used by the wire, webauthn, debug, and vcp interfaces.

This comment in the usb.py file confuses me: "We only have 10 available USB endpoints on real HW, of which 2 are taken up by the USB descriptor." I'm guessing that there is a mis-config/bug at a lower level that caused this observation to seem accurate.

The only place that I have found any mention of this endpoint restriction is in some STM HAL code (when searching for `USE_USB_HS_IN_FS`). There is no explanation why. Nothing in the manuals say that not all endpoints are available when using the HS with the FS PHY. The FIFO data RAM does not disappear.

I think that the `USE_USB_HS_IN_FS` restriction on `dev_endpoints` could be a held-over bug. The current uPython code looks like it agrees with my hypothesis:
https://github.com/micropython/micropython/blob/master/ports/stm32/usbd_conf.c#L427

`MICROPY_HW_USB_HS_IN_FS` does not concern itself with the number of available endpoints (at least, not anymore).

https://github.com/micropython/micropython/blob/master/ports/stm32/boards/STM32F429DISC/mpconfigboard.h#L75

**Comment by mcudev** [comment]
I have some changes to core/embed/trezorhal/usbd_conf.c and core/src/usb.py that let me run everything all at once on an actual device with no nasty kernel messages (and lsusb shows all the config correctly). I'll push something up later to look at. I'm going to take a break and look at it again to make sure that I'm not somehow fooling myself (feeling a little burned out right now).

**Comment by mcudev** [comment]
Here's something that gets all the interfaces working at the same time on a TT device with a debug firmware.
https://github.com/trezor/trezor-firmware/compare/master...mcudev:usb_endpoints

The ST documentation on the USB FIFO is pretty horrible, so I'm not sure yet if this is optimal or even all that stable.
It's definitely not ready to propose for merging. But, it's a step further and seems to (mostly) work.
Most of what's interesting is in `core/embed/trezorhal/usbd_conf.c`.

Here are some helpful commands and notes for testing:

`PYOPT=0 make build_firmware`
`make upload`

`screen /dev/ttyACM0`
`trezorctl list`
`trezorctl device setup`
`trezorctl set pin`

test fido with the chrome dev tools script:

```
navigator.credentials.create({
  publicKey: {
  "rp": {
    "name": "Gemini Trust Company, LLC",
    "id": "gemini.com"
  },
  user: {
    id: new Uint8Array([1]),
    name: "user@email.test",
    displayName: "whatever"
  },
  challenge: new Uint8Array([1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]),
  "pubKeyCredParams": [{"alg": -7, "type": "public-key"}, {"alg": -257, "type": "public-key"}],
  "authenticatorSelection": {"requireResidentKey": false, "userVerification": "preferred"},
  "attestation": "none",
  "extensions":{}
}
});
```

check kernel messages with: `journalctl -ke`
check usb config with: `lsusb --verbose -d 1209:`

```
Bus 003 Device 054: ID 1209:53c1 Generic SatoshiLabs TREZOR
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.10
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1209 Generic
  idProduct          0x53c1 SatoshiLabs TREZOR
  bcdDevice            2.00
  iManufacturer           1 SatoshiLabs
  iProduct                2 TREZOR
  iSerial                 3 6B83BE4ACCC545E0A9DB6BB0
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0099
    bNumInterfaces          5
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0x80
      (Bus Powered)
    MaxPower              100mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           2
      bInterfaceClass       255 Vendor Specific Class
      bInterfaceSubClass      0 
      bInterfaceProtocol      0 
      iInterface              5 TREZOR Interface
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x01  EP 1 OUT
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           2
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      0 
      iInterface              5 TREZOR Interface
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      34
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x02  EP 2 OUT
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           2
      bInterfaceClass       255 Vendor Specific Class
      bInterfaceSubClass      0 
      bInterfaceProtocol      0 
      iInterface              5 TREZOR Interface
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x03  EP 3 OUT
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
    Interface Association:
      bLength                 8
      bDescriptorType        11
      bFirstInterface         3
      bInterfaceCount         2
      bFunctionClass          2 Communications
      bFunctionSubClass       2 Abstract (modem)
      bFunctionProtocol       1 AT-commands (v.25ter)
      iFunction               0 
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        3
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         2 Communications
      bInterfaceSubClass      2 Abstract (modem)
      bInterfaceProtocol      1 AT-commands (v.25ter)
      iInterface              5 TREZOR Interface
      CDC Header:
        bcdCDC               10.01
      CDC Call Management:
        bmCapabilities       0x00
        bDataInterface          4
      CDC ACM:
        bmCapabilities       0x02
          line coding and serial state
      CDC Union:
        bMasterInterface        3
        bSlaveInterface         4 
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x85  EP 5 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval              10
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        4
      bAlternateSetting       0
      bNumEndpoints           2
      bInterfaceClass        10 CDC Data
      bInterfaceSubClass      0 
      bInterfaceProtocol      0 
      iInterface              5 TREZOR Interface
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x84  EP 4 IN
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               0
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x04  EP 4 OUT
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               0
Binary Object Store Descriptor:
  bLength                 5
  bDescriptorType        15
  wTotalLength       0x001d
  bNumDeviceCaps          1
  Platform Device Capability:
    bLength                24
    bDescriptorType        16
    bDevCapabilityType      5
    bReserved               0
    PlatformCapabilityUUID    {3408b638-09a9-47a0-8bfd-a0768815b665}
      WebUSB:
        bcdVersion    1.00
        bVendorCode      1
        iLandingPage     0 
can't get debug descriptor: Resource temporarily unavailable
Device Status:     0x0000
  (Bus Powered)
```

**Comment by mcudev** [comment]
I found some more info on the FIFO sizing and I updated accordingly https://github.com/trezor/trezor-firmware/compare/master...mcudev:usb_endpoints
Seems to work nicely. With these new configs, I haven't encountered any snags/slowdowns/etc...

**Comment by mcudev** [comment]
Tested ACK.

I verified that `./emu.py` uses UDP ports 127.0.0.1:21324, 127.0.0.1:21325, and 127.0.0.1:21326 and that hid-bridge works with it.

I verified that `./emu.py -d` uses UDP ports 127.0.0.1:21324 and 127.0.0.1:21326 and that hid-bridge works with it.

I tested trezorctl wire interface commands and FIDO2 registrations with chrome against the emulator in both configurations. The `chrome://device-log/` looked clean too.

I think this is ok to merge and then we can work #1517 on top of this.

**Review by prusnak — APPROVED** [review]
